### PR TITLE
Add identification of golang binaries

### PR DESCRIPTION
**Pull request by marcin-ol**
Support identification of Golang binaries by looking for ELF section `.note.go.buildid`, supported since Go 1.5.
Reference: https://github.com/golang/go/issues/11048.

### CODE REVIEW

File: granulate_utils/golang.py [hunk 1]
```diff
@@ -9,12 +9,26 @@
...
+@functools.lru_cache(maxsize=4096)
+def get_golang_buildid(process: Process) -> Optional[str]:
+    """
+    Gets the golang build ID embedded in an ELF file section as a string, or None if not present.
+    """
+    elf_path = f"/proc/{process.pid}/exe"
+    try:
+        # section .note.go.buildid has been added since version 1.5: https://github.com/golang/go/issues/11048
+        golang_buildid = get_elf_buildid(elf_path, ".note.go.buildid", lambda note: note.n_name == "Go")
+        return golang_buildid
```
Comment (Jongy) [inline granulate_utils/golang.py:29]:
```suggestion
        return get_elf_buildid(elf_path, ".note.go.buildid", lambda note: note.n_name == "Go")
```

File: granulate_utils/linux/elf.py [hunk 2]
```diff
@@ -71,18 +71,19 @@ def get_elf_arch(elf: ELFType) -> str:
         return elf.get_machine_arch()
 
 
-def get_elf_buildid(elf: ELFType) -> Optional[str]:
+def get_elf_buildid(elf: ELFType, section: str, note_check: Callable[[Any], Any] = lambda n: True) -> Optional[str]:
```
Comment (Jongy) [inline granulate_utils/linux/elf.py:74]: Does it make sense to have a default value for the note? And furthermore, a default value that behaves differently that the past behavior?

I'd just remove the default and require the caller to pass it.

If you can add types that's also good :)

**Comment by Jongy** [comment]
Related: https://github.com/Granulate/gprofiler/issues/817

File: granulate_utils/golang.py (hunk 1 above)
Comment (marcin-ol) [reply granulate_utils/golang.py]: 👍

File: granulate_utils/linux/elf.py (hunk 2 above)
Comment (marcin-ol) [reply granulate_utils/linux/elf.py]: 👍

**Comment by Jongy** [comment]
Btw this change will have a side effect of greatly improving the performance (iterating over sections is much less work than iterating over symbols), helping for https://github.com/Granulate/gprofiler/issues/696 cause.

File: granulate_utils/linux/elf.py [hunk 3]
```diff
@@ -71,18 +71,19 @@ def get_elf_arch(elf: ELFType) -> str:
...
+def get_elf_buildid(elf: ELFType, section: str, note_check: Callable[[NoteSection], bool]) -> Optional[str]:
     """
-    Gets the build ID embedded in an ELF file section as a hex string,
+    Gets the build ID embedded in an ELF file note section as a hex string,
     or None if not present.
+    Lambda argument is used to verify that note meets caller's requirements.
     """
     with open_elf(elf) as elf:
-        build_id_section = elf.get_section_by_name(".note.gnu.build-id")
+        build_id_section = elf.get_section_by_name(section)
```
Comment (Jongy) [inline granulate_utils/linux/elf.py:81]: Can name it note_section as it's not necessarily a build-id one anymore

**Review by Jongy — CHANGES_REQUESTED** [review]
One final small comment.

Other than that I'll have to review the usage of changed functions (`is_golang_process` and `get_elf_buildid`) in other project which use granulate-utils to ensure compatibility is kept.

**Comment by Jongy** [comment]
No problematic usages. I'm ready to merge this - just waiting on your reply on the last comment.

**Review by Jongy — APPROVED** [review]

**Pull request merged**